File: 12. Error Handling and Building Fault Tolerant Systems/code/typescript/③_safe_logging_practices_4.ts
// TypeScript, Safe logging practices
// Assuming a configured logger like winston or pino
import pino from 'pino';
const logger = pino();

// X UNSAFE, never do this
/*
logger.error({
    email: user.Email,          // PII leak
    password: req.Password,     // catastrophic
    api_key: cfg.OpenAIKey,     // secret leak
}, "login_failed");
*/

// OK SAFE, IDs and correlation only
/*
logger.error({
    user_id: user.ID,
    correlation_id: req.headers['x-request-id'],
    reason: "invalid_credentials",  // generic code, not DB message
}, "login_failed");
*/

File: 12. Error Handling and Building Fault Tolerant Systems/code/rust/③_safe_logging_practices_4.rs
// Rust, Safe logging practices
use log::error;

// X UNSAFE, never do this
/*
error!(
    "login_failed email={} password={} api_key={}",
    user.email,         // PII leak
    req.password,       // catastrophic
    cfg.openai_key      // secret leak
);
*/

// OK SAFE, IDs and correlation only
/*
error!(
    "login_failed user_id={} correlation_id={} reason={}",
    user.id,
    req.headers().get("X-Request-ID").unwrap_or(""),
    "invalid_credentials" // generic code, not DB message
);
*/
